include/nbl/video/IGPUPipeline.h

```diff
@@ -0,0 +1,110 @@
+                // https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineShaderStageCreateInfo.html#VUID-VkPipelineShaderStageCreateInfo-maxClipDistances-00708
+                // to:
+                // https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineShaderStageCreateInfo.html#VUID-VkPipelineShaderStageCreateInfo-stage-06686
+                // and from:
+                // https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineShaderStageCreateInfo.html#VUID-VkPipelineShaderStageCreateInfo-pNext-02756
+                // to:
+                // https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineShaderStageCreateInfo.html#VUID-VkPipelineShaderStageCreateInfo-module-08987
+                  
+                int64_t specData = 0;
+                for (const auto& entry : *entries)
+                {
+                  if (!entry.second)
+                      return INVALID_SPEC_INFO;
+                  specData += entry.second.size();
+                }
+                if (specData>0x7fffffff)
+                    return INVALID_SPEC_INFO;
+                return static_cast<int32_t>(specData);
+            }
+
+            const asset::IShader* shader = nullptr;
+            std::string_view entryPoint = "";
+            asset::IPipelineBase::SUBGROUP_SIZE requiredSubgroupSize : 3 = asset::IPipelineBase::SUBGROUP_SIZE::UNKNOWN;	//!< Default value of 8 means no requirement
+            uint8_t requireFullSubgroups : 1 = false;
+
+            // Container choice implicitly satisfies:
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkSpecializationInfo.html#VUID-VkSpecializationInfo-constantID-04911
+            const core::unordered_map<spec_constant_id_t, SSpecConstantValue>* entries;
+            // By requiring Nabla Core Profile features we implicitly satisfy:
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineShaderStageCreateInfo.html#VUID-VkPipelineShaderStageCreateInfo-flags-02784
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineShaderStageCreateInfo.html#VUID-VkPipelineShaderStageCreateInfo-flags-02785
+            // Also because our API is sane, it satisfies the following by construction:
+            // https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineShaderStageCreateInfo.html#VUID-VkPipelineShaderStageCreateInfo-pNext-02754
+
+        };
+
+};
+
+// Common Base class for pipelines
+template<typename PipelineNonAssetBase>
+class IGPUPipeline : public IBackendObject, public PipelineNonAssetBase, public IGPUPipelineBase
+{
+    protected:
+
+        template <typename... Args>
+        explicit IGPUPipeline(core::smart_refctd_ptr<const ILogicalDevice>&& device, Args&&... args) :
+         PipelineNonAssetBase(std::forward<Args>(args...)), IBackendObject(std::move(device))
+        {}
```

Comment:
no need for perfect forwarding, you know that `PipelineNonAssetBase` is `IPipeline<IGPUPipelineLayout>` and it only takes a pipeline layout smartpointer